va/va.h

```diff
@@ -129,6 +129,99 @@ extern "C" {
  * 	- \ref api_fei
  * 	- \ref api_fei_h264
  * 	- \ref api_fei_hevc
+ * 
+ * \section threading Multithreading Guide
+ * All VAAPI functions implemented in libva should be thread-safe. However, 
+ * VAAPI and libva do not impose thread-safety requirements on the driver or
+ * the backend implementation, and it is the responsibility of the driver or the
+ * backend implementation to ensure thread-saftey of VAAPI functions. In other 
+ * words, VAAPI functions are thread-safe only if the driver or the backend 
+ * implementation is also thread-safe. 
+ *
+ * It is assumed that none of the VAAPI functions will be called from signal 
+ * handlers.
+ *
+ * Thread-safety in this context means that when VAAPI is being called by multiple
+ * concurrent threads, it will not crash or hang the OS, and VAAPI internal
+ * states will not be corrupted. When multiple threads are operating on the same
+ * VAAPI objects, it is the application's responsibility to synchronize these
+ * operations in order to generate the expected results.
+ *
+ * It is worth noting that VAContext is not meant to be shared among different 
+ * threads. i.e if a VAContext object is shared among two or more threads, its 
+ * internal state may get corrupted which may lead to unexpected results for 
+ * operations using this VAContext. It is expected that each thread shall create
+ * and maintain its own VAContext object for all operations.  
```

Comment:
> VAContext is not meant to be shared among different threads

I believe this is a too unclear statement. "Is not meant (or intended)" can be read as "could be and this is not prohibited according to the API". That's how I read it at least.

If we will transform that to the strict requirement, then I believe it is too strong requirement. I don't actually see why single VAContext can't be used from few threads. For example, it can be created in one thread, tasks submitted in another thread and it can be destroyed in yet another thread. What's wrong in such usage and why this can't be done as soon as operations are serialized?

I would suggested to be more specific and clear. Maybe we should better clarify whether it is application's responsibility to serialize vaBeginFrame-vaEndFrame sequencies? I.e. you can't have 2 concurrent vaBeginFrame-vaEndPicture sequences at a time. Having a single sequence serialized, but executed in different threads is likely OK.

Reply:
Updated based on the feedbacks